cabal-install/src/Distribution/Client/Config.hs

```diff
@@ -587,40 +590,87 @@ initialSavedConfig = do
     }
   }
 
-defaultCabalDir :: IO FilePath
-defaultCabalDir = getAppUserDataDirectory "cabal"
-
-getCabalDir :: IO FilePath
-getCabalDir = do
+-- | If @CABAL\_DIR@ is set or @~/.cabal@ exists, return that
+-- directory.
+maybeGetCabalDir :: IO (Maybe FilePath)
+maybeGetCabalDir = do
   mDir <- lookupEnv "CABAL_DIR"
   case mDir of
-    Nothing -> defaultCabalDir
-    Just dir -> return dir
+    Just dir -> return $ Just dir
+    Nothing -> do
+      defaultDir <- getAppUserDataDirectory "cabal"
+      dotCabalExists <- doesDirectoryExist defaultDir
+      return $ if dotCabalExists
+               then Just defaultDir
+               else Nothing
+
+-- The default behaviour of cabal-install is to use the XDG directory
```

Comment:
Any reason why this is not a proper Haddock declaration? :)

Reply:
I have the habit of not using Haddock declarations for non-exported functions.  I vaguely recall that there was a good reason for this, but I don't remember it anymore, so I will change this.